compiler/rustc_mir_transform/src/ssa.rs

```diff
@@ -19,6 +19,33 @@ pub struct SsaLocals {
     copy_classes: IndexVec<Local, Local>,
 }
 
+/// We often encounter MIR bodies with 1 or 2 basic blocks. In those cases, it's unnecessary to
+/// actually compute dominators, we can just compare block indices because bb0 is always the first
+/// block, and in any body all other blocks are always always dominated by bb0.
```

Comment:
If there are 1 or 2 bbs, dominators are probably trivial.
But this idea is very interesting if the CFG is a tree, ie if all bbs have a single incoming edge.

Reply:
We could also cache the dominators to avoid recomputations when cfg does not change. tmiasko had a recent PR doing this.

Reply:
I'm not entirely sure what you're getting at. I just wrote this based on the fact that the `derive` benchmark that this currently regresses only contains functions with one or two basic blocks, and even there the general-purpose dominators calculation is ~22% of CopyProp's runtime.

I can see how this is easily extended to CFGs that are a trivial chain of basic blocks, but I don't see how to extend this to trees. Surely we still need to store some state there? Or does something about the traversal save us?